### PR TITLE
Only hyphenate text for which there is a language specified

### PR DESCRIPTION
**Pull request by obyknovenius**
#### 32780e4c841e62600407aea0f1ec6a6471b22e94
<pre>
Only hyphenate text for which there is a language specified
<a href="https://bugs.webkit.org/show_bug.cgi?id=193003">https://bugs.webkit.org/show_bug.cgi?id=193003</a>

Reviewed by Alex Christensen.

The spec <a href="https://drafts.csswg.org/css-text/#hyphens-property">https://drafts.csswg.org/css-text/#hyphens-property</a> says:

&quot;Authors should correctly tag their content’s language (e.g. using the
HTML lang attribute or XML xml:lang attribute) in order to obtain
correct automatic hyphenation.&quot;

There&apos;s also a dedicated WPT for this:
`imported/w3c/web-platform-tests/css/css-text/hyphens/hyphens-auto-001.html`

This patch makes WebKit follow the spec.

* LayoutTests/TestExpectations:
* LayoutTests/editing/spelling/spelling-marker-includes-hyphen-expected.html:
* LayoutTests/editing/spelling/spelling-marker-includes-hyphen.html:
* LayoutTests/fast/inline/hyphenation-when-overflow-wrap-is-break-word-expected.html:
* LayoutTests/fast/inline/hyphenation-when-overflow-wrap-is-break-word.html:
* LayoutTests/fast/text/simple-line-hyphens-with-text-align.html:
* LayoutTests/fast/text/simple-line-hyphens-with-word-letter-spacing.html:
* LayoutTests/fast/text/simple-line-layout-hyphen-limit-after.html:
* LayoutTests/fast/text/simple-line-layout-hyphen-limit-before.html:
* LayoutTests/fast/text/simple-line-layout-hyphen-limit-lines.html:
* LayoutTests/fast/text/simple-line-layout-hyphenation-constrains.html:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/gtk/fast/text/hyphenate-locale-expected.png:
* LayoutTests/platform/gtk/fast/text/hyphenate-locale-expected.txt:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/mac/fast/text/hyphenate-locale-expected.png:
* LayoutTests/platform/mac/fast/text/hyphenate-locale-expected.txt:
* LayoutTests/platform/wpe/fast/inline/hyphenation-when-overflow-wrap-is-break-word-expected.html: Added.
* LayoutTests/platform/wpe/fast/text/simple-line-layout-hyphenation-constrains-expected.html: Copied from LayoutTests/fast/text/simple-line-layout-hyphenation-constrains.html.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-text/hyphens/hyphenate-character-002-expected.html: Added.
* Source/WebCore/platform/text/cf/HyphenationCF.cpp:
(WebCore::canHyphenate):
* Source/WebCore/platform/text/hyphen/HyphenationLibHyphen.cpp:
(WebCore::canHyphenate):
(WebCore::lastHyphenLocation):

Canonical link: <a href="https://commits.webkit.org/284355@main">https://commits.webkit.org/284355@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/3d3fb29d377241ef6d89c5c260f1dbf5e1afa1db

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69134 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48534 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21806 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73215 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20292 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71251 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56335 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20141 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55021 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13471 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72200 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44299 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59682 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35500 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40967 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17112 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18666 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62910 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17457 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74926 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13116 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16696 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62676 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13154 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59765 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62576 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15341 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10576 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4184 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44338 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45411 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46607 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45153 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/platform/text/cf/HyphenationCF.cpp [hunk 1]
```diff
@@ -51,10 +51,7 @@ class TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>
 
     static RetainPtr<CFLocaleRef> createValueForNullKey()
```
Comment (achristensen07) [inline Source/WebCore/platform/text/cf/HyphenationCF.cpp:52]: If this is what we want to do, we should remove this function.

**Review by achristensen07 — COMMENTED** [review]
Looks like another test needs fixing.

File: LayoutTests/platform/wpe/fast/inline/hyphenation-when-overflow-wrap-is-break-word-expected.html [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<style>
+div {
+  font-size: 16px;
+  border: 1px solid green;
+  width: 50px;
+  -webkit-hyphens: auto;
+}
+</style>
+<div lang="en">Autom<br>atic<br>hyphen<br>ation<br>opportu<br>nities<br>overrid<br>e<br>arbitrar<br>y<br>position<br>s</div>
```
Comment (rniwa) [inline LayoutTests/platform/wpe/fast/inline/hyphenation-when-overflow-wrap-is-break-word-expected.html:9]: Why is this file in platform/wpe? It should be in the same directory as the test.

File: Source/WebCore/platform/text/cf/HyphenationCF.cpp [hunk 3]
```diff
@@ -35,49 +35,29 @@
 namespace WTF {
 
 template<>
-class TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>
+RetainPtr<CFLocaleRef> TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>::createValueForKey(const AtomString& localeIdentifier)
 {
-public:
-    static TinyLRUCache<AtomString, RetainPtr<CFLocaleRef>>& cache()
```
Comment (rniwa) [inline Source/WebCore/platform/text/cf/HyphenationCF.cpp]: What's up with all these unrelated code getting removed?

File: Source/WebCore/platform/text/cf/HyphenationCF.cpp [hunk 4]
```diff
@@ -35,49 +35,29 @@
...
 
+static RetainPtr<CFLocaleRef> cachedLocale(const AtomString& localeIdentifier)
+{
+    static NeverDestroyed<TinyLRUCache<AtomString, RetainPtr<CFLocaleRef>>> cache;
+    return cache.get().get(localeIdentifier);
+}
 bool canHyphenate(const AtomString& localeIdentifier)
 {
-    return TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>::cache().get(localeIdentifier);
+    return cachedLocale(localeIdentifier);
```
Comment (rniwa) [inline Source/WebCore/platform/text/cf/HyphenationCF.cpp]: This change seems unnecessary.

File: LayoutTests/platform/wpe/fast/inline/hyphenation-when-overflow-wrap-is-break-word-expected.html (hunk 2 above)
Comment (obyknovenius) [reply LayoutTests/platform/wpe/fast/inline/hyphenation-when-overflow-wrap-is-break-word-expected.html:9]: WPE doesn't support hyphenation. That's why it requires a separate baseline. Other ports use `LayoutTests/fast/inline/hyphenation-when-overflow-wrap-is-break-word-expected.html` from the same directory as the test.

File: Source/WebCore/platform/text/cf/HyphenationCF.cpp (hunk 3 above)
Comment (obyknovenius) [reply Source/WebCore/platform/text/cf/HyphenationCF.cpp]: It's related. As @achristensen07 commented, since we no longer fall back to the default locale when the `localeIdentifier` is null, we don't need to implement `TinyLRUCache::isKeyNull()` and `TinyLRUCache::createValueForNullKey()`. Thus this code can be simplified to the implementation of `TinyLRUCache::createValueForKey()`.

File: Source/WebCore/platform/text/cf/HyphenationCF.cpp (hunk 4 above)
Comment (obyknovenius) [reply Source/WebCore/platform/text/cf/HyphenationCF.cpp]: Since the templated `TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>` class was removed, we can't use the `static TinyLRUCache<AtomString, RetainPtr<CFLocaleRef>>& cache()` method anymore. A new static `cachedLocale()` function was added to create and access the` NeverDestroyed<TinyLRUCache<AtomString, RetainPtr<CFLocaleRef>>>` cache. It's also shorter.

Comment (vitorroriz) [reply Source/WebCore/platform/text/cf/HyphenationCF.cpp]: Does the refactoring change behavior per se? Otherwise, maybe we should split this into 2 patches. One for refactoring and one for addressing the hyphenation bug, as it would be easier to follow.

Comment (rniwa) [reply Source/WebCore/platform/text/cf/HyphenationCF.cpp]: Why are we removing `TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>`? I don't think we should do that. It makes PR impossible to follow.

Comment (obyknovenius) [reply Source/WebCore/platform/text/cf/HyphenationCF.cpp]: Note that `TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>` is a **class** template specialization. If we keep using it we have to implement all 4 static methods, where 3 of them are almost identical to the templated implementation. Also, `isKeyNull()` and `createValueForNullKey()` are redundant since we don't store a locale for null `localeIdentifier`. It would look something like this.

```cpp
namespace WTF {

template<>
class TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>
{
public:
    static TinyLRUCache<AtomString, RetainPtr<CFLocaleRef>>& cache()
    {
        static NeverDestroyed<TinyLRUCache<AtomString, RetainPtr<CFLocaleRef>>> cache;
        return cache;
    }

    static bool isKeyNull(const AtomString& localeIdentifier)
    {
        return localeIdentifier.isNull();
    }

    static RetainPtr<CFLocaleRef> createValueForNullKey()
    {
        return nullptr;
    }

    static RetainPtr<CFLocaleRef> createValueForKey(const AtomString& localeIdentifier)
    {
        RetainPtr<CFLocaleRef> locale = adoptCF(CFLocaleCreate(kCFAllocatorDefault, localeIdentifier.string().createCFString().get()));

        return CFStringIsHyphenationAvailableForLocale(locale.get()) ? locale : nullptr;
    }

    static AtomString createKeyForStorage(const AtomString& key) { return key; }
};

} // namespace WTF
```
What I propose instead is to use a **function** template specialization and implement only `createValueForKey()` method:
```cpp
namespace WTF {

template<>
RetainPtr<CFLocaleRef> TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>::createValueForKey(const AtomString& localeIdentifier)
{
    RetainPtr<CFLocaleRef> locale = adoptCF(CFLocaleCreate(kCFAllocatorDefault, localeIdentifier.string().createCFString().get()));
    return CFStringIsHyphenationAvailableForLocale(locale.get()) ? locale : nullptr;
}

} // namespace WTF

namespace WebCore {

static RetainPtr<CFLocaleRef> cachedLocale(const AtomString& localeIdentifier)
{
    static NeverDestroyed<TinyLRUCache<AtomString, RetainPtr<CFLocaleRef>>> cache;
    return cache.get().get(localeIdentifier);
}

} // namespace WebCore
```

It's more compact and less confusing in my opinion.

This is similar to the `CGColor` cache in https://github.com/WebKit/WebKit/blob/e89434de872e7aa88c2d2119df0e30b1b8004d9e/Source/WebCore/platform/graphics/cg/ColorCG.cpp#L45-L53

File: LayoutTests/platform/mac/TestExpectations [hunk 5]
```diff
@@ -2425,3 +2425,10 @@ imported/w3c/web-platform-tests/css/css-writing-modes/text-orientation-upright-s
 [ Sonoma+ ] http/tests/media/video-error-abort.html [ Failure ]
 [ Sonoma+ ] http/tests/media/video-served-as-text.html [ Failure ]
 [ Sonoma+ ] http/tests/media/video-throttled-load-metadata.html [ Failure ]
+# webkit.org/b/277720 [ iOS MacOS ] 4x http/wpt/service-workers/fetch-service-worker-preload* (layout-tests) are flaky failures
```
Comment (achristensen07) [inline LayoutTests/platform/mac/TestExpectations]: This change looks unrelated and should be removed from this PR.

**Review by achristensen07 — APPROVED** [review]
